pkg/plugin/loader.go

```diff
@@ -82,14 +87,59 @@ func checkDirAndCompile(pluginType, driver string) (string, error) {
 	return bin, nil
 }
 
+// downloadPlugin downloads the plugin and stores it into local filesystem
+func downloadAndCompilePlugin(pluginType, driver string) (string, error) {
+	driverURL := strings.TrimPrefix(driver, "https://")
```

Comment:
This is needed because of go-getter. It errors when url is of the type `https://github.com/jimil749/...`, it should be of the type: `github.com/jimil749/...`

Reply:
This isn't really good. We need to support other protocols such as git as well. It even says there in the README.

> Protocols are used to download files/directories using a specific mechanism. Example protocols are Git and HTTP.

We can pass the getters to the client https://github.com/hashicorp/go-getter/blob/main/client.go#L144 https://github.com/hashicorp/go-getter/blob/main/get.go#L67-L75

Reply:
With the current mechanism, we are supporting Git protocol. The URL of the type `github.com/jimil749/json` would be turned into Git URL by go-getter. The go-getter uses Git protocol to fetch the repo.

> The concept of a detector automatically turns invalid URLs into proper URLs. For example: "github.com/hashicorp/go-getter" would turn into a Git URL. Or "./foo" would turn into a file URL

Reason to trim `https://` was because `github.com/jimil/...` is not a valid URL (w/o the https:// part), hence the url checker would fail if one entered URL of that kind.

We could get rid of this by changing the way we check the URL.

Reply:
The user can pass paths beginning with other protocols, eg. http, git, file. We'll have to trim all these separately. Let's try the approach in the above comment

Reply:
Ahhh, so we're supporting protocols other than git! User can just add a prefix as per their choices.

Reply:
Also, I didn't go through how the library's clients store the transport, but let's ensure that we don't leak fds.

Reply:
So @ishank011, I don't think we'd need to pass our own Getter, because all the protocols like git, http, file are supported out of the box and are detected based on the `Src` field. For instance, if the url is `github.com/abc/xyz` -> git protocol or if url is `./foo` -> file protocol. This is done by the go-getter client. We just need to pass the url and rest would be taken care by the `getter.Get()` method!

Reply:
I didn't mean writing our own getter implementation, but just passing the in-built ones when instantiating the client. The problem with the current method is that we need to anticipate the protocols and trim all of those. If we didn't have to do that, then it would have been perfectly fine. So rather than the trimming, I'd prefer that we just make it known to the client.

Reply:
We don't actually need to pass the in-built getter because by default `go-getter` already uses those if not specified in the client (https://github.com/hashicorp/go-getter/blob/main/client_option.go#L29). So if we need to support multiple protocols and also not use trimming, we could ask the users to specify urls like: `git::https://github.com/repo/plugin` which would force the protocol desired by the user.

Reply:
Doesn't look like a good option to me. We ask the use to specify the protocol, and then specify another to override it. I'd still prefer passing the getters at init.

Reply:
Sure, just want to know that how would the user configure the plugins then. Could you provide an example of diff types of URLs/Protocol configuration?

Reply:
Sure so
```
https://github.com/repo/plugin
http://github.com/repo/plugin
s3://my-s3-deployment.com/bucket/repo/plugin
file:///root/repo/plugin.git
```

Reply:
Ok, so the only ambiguity is b/w `git` and `http` protocol. Even if I pass the getters as mentioned [here](https://github.com/hashicorp/go-getter/blob/main/get.go#L67-L75) (which I think we don't need to because those are used by default if no getter is passed while instantiating the client), there would be ambiguity b/w git and http protocol. The correct way to use `git` protocol is by passing url like `github.com/jimil749/json` and not `http://github.com/jimil749/json` (wherein the getter would use http protocol). Sorry, but I am still confused as in what would passing the getter to Client achieve?